crates/hulk/src/nao/double_buffered_reader.rs

```diff
@@ -0,0 +1,411 @@
+use std::{
+    io::{self, ErrorKind, Read},
+    mem::size_of,
+    os::unix::{io::AsRawFd, prelude::RawFd},
+};
+
+use super::hula::StateStorage;
+
+pub struct DoubleBufferedReader {
+    buffers: [[u8; size_of::<StateStorage>()]; 2],
+    active_buffer_index: usize,
+    incomplete_offset: usize,
+}
+
+impl Default for DoubleBufferedReader {
+    fn default() -> Self {
+        Self {
+            buffers: [[0; size_of::<StateStorage>()]; 2],
+            active_buffer_index: Default::default(),
+            incomplete_offset: Default::default(),
+        }
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
+    }
+
+    pub fn read(
+        &mut self,
+        reader: &mut (impl AsRawFd + Read),
+        mut poll_reader: impl FnMut(RawFd) -> io::Result<()>,
+    ) -> io::Result<()> {
+        let mut buffers_swapped = false;
+        loop {
+            match reader.read(&mut self.buffers[self.active_buffer_index][self.incomplete_offset..])
+            {
+                Ok(number_of_read_bytes) => {
+                    self.incomplete_offset += number_of_read_bytes;
+                    assert!(self.incomplete_offset <= self.buffers[self.active_buffer_index].len());
+                    if self.incomplete_offset == self.buffers[self.active_buffer_index].len() {
+                        self.swap_buffers();
+                        buffers_swapped = true;
+                    }
+                }
+                Err(ref error) if error.kind() == ErrorKind::WouldBlock => {
+                    if buffers_swapped {
+                        return Ok(());
+                    }
+                    poll_reader(reader.as_raw_fd())?;
+                }
+                Err(error) => return Err(error),
+            }
+        }
+    }
+
+    /// Precondition: `read()` has been called once
+    pub fn get_last(&self) -> &StateStorage {
```

Comment:
I'm not a fan of `get_[...]` maybe just drop the `get`-part?

Reply:
Another suggestion is `buffered()`? Because `last` indicates, there is more to get, but the user can actually always only get one complete message out of the buffer. So there is no first or last, there is just one `buffered` message.

I'm open to for suggestions on the actual name, but i would like to see this concept represented.

Reply:
The `DoubleBufferingReader::drain()` now returns the read value directly. The precondition is also enforced by that design. `comments -= 1`